### DISCUSSION TITLE
The google gemini api gives an error when used in litellm.

### DISCUSSION THREAD

**Discussion by flyfox666** [open-webui/open-webui#1482] [General]
# Bug Report

## Description

**Bug Summary:**
The google gemini api gives an error when used in litellm，when use in chat, error show:
GenerationConfig.__init__() got an unexpected keyword argument 'num_ctx'

**Steps to Reproduce:**
add gemini api in webui ,then refresh ,then use the model to chat

**Expected Behavior:**
no error

**Actual Behavior:**
error showen

## Environment

- **Operating System:** win11 ,ubuntu 22.04, docker for windows
- **Browser (if applicable):**chrome latest
- 
## Reproduction Details

**Confirmation:**

- [x] I have read and followed all the instructions provided in the README.md.
- [x] I am on the latest version of both Open WebUI and Ollama.
- [x] I have included the browser console logs.
- [x] I have included the Docker container logs.

## Logs and Screenshots

**Docker Container Logs:**
[Include relevant Docker container logs, if applicable]
02:16:19 - LiteLLM Router:INFO: litellm.acompletion(model=gemini/gemini-pro) Exception GenerationConfig.__init__() got an unexpected keyword argument 'num_ctx'
INFO:LiteLLM Router:litellm.acompletion(model=gemini/gemini-pro) Exception GenerationConfig.__init__() got an unexpected keyword argument 'num_ctx'
Traceback (most recent call last):
  File "/usr/local/lib/python3.11/site-packages/litellm/llms/gemini.py", line 177, in async_streaming
    generation_config=genai.types.GenerationConfig(
                      ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
TypeError: GenerationConfig.__init__() got an unexpected keyword argument 'num_ctx'
During handling of the above exception, another exception occurred:
Traceback (most recent call last):
  File "/usr/local/lib/python3.11/site-packages/litellm/main.py", line 279, in acompletion
    response = await init_response
               ^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/litellm/llms/gemini.py", line 192, in async_streaming
    raise GeminiError(status_code=500, message=str(e))

![gemini pro api](https://github.com/open-webui/open-webui/assets/121539277/b8d16c57-eace-41d1-83df-62b21b0f215e)

**Screenshots (if applicable):**
[Attach any relevant screenshots to help illustrate the issue]

**Comment by tjbck** [comment]
Could you verify that you didn't have any values for advanced params in settings? Keep us updated!

**Reply by flyfox666** [reply]
HI tjbck
attached the setting
![111](https://github.com/open-webui/open-webui/assets/121539277/9a10e50d-ab2d-453a-a593-d4c26c0deb47)

I just reset all the docker and reinstall, still not work.

Follow the gemini yaml data 

- litellm_params:
    api_key: xxxxx
    model: gemini/gemini-pro
  model_info:
    id: e55829f9-e0a1-4e44-a6db-f067ba56f2ef
  model_name: gemini/gemini-pro

**Comment by flyfox666** [comment]
@tjbck the problem just suddenly sovled , I just clear the cookie in chrome. And then every thing ok..

**Comment by justinh-rahb** [comment]
@tjbck is `num_ctx` not an Ollama-specific variable name? afaik OpenAI API's don't support this parameter so why would we even send it?

**Reply by rob-x-ai** [reply]
Also when calling Anthropic's API I have a similar error when I have parameters set.

**Reply by flyfox666** [reply]
Try to have a look at https://github.com/BerriAI/litellm/blob/main/litellm/llms/anthropic.py

add sth before line 398 ##logging , i upload the py file to gpt4 let it give me some orders base the gemini.py rule

**Comment by flyfox666** [comment]
HI @tjbck @justinh-rahb 
     Last night I tested a couple rounds with elimination and found out under what circumstances this problem occurs as follows:
1. Modify the context length or max tokens option in Settings---General---Advanced Parameters, e.g. adjust the context length or max tokens up or down, and save it. 
2. Select the gemin-pro model set in litemllm for dialog, then GenerationConfig.__init__() got an unexpected keyword argument 'num_ctx'. 
3. After reverting the changes in step 1 to the original values, the same error still occurs. 
Tried solution 1: clear chrome's cookies, log in again, and select the geminin-pro model set in litemllm to have a conversation. No error happend
Tried solution 2: open a new traceless browser and log in.No error happend

**Reply by justinh-rahb** [reply]
Ya, that's consistent with what I thought, this parameter (that is relevant to a runtime like Ollama that can use many different models) is not supported by external APIs and shouldn't be being sent in requests. Possible that some APIs just ignore it and pass the request anyway, but it seems that Google is more strict about invalid requests. Where it might be complicated or confused here is that LiteLLM can use providers that _do_ support this parameter, so disabling it altogether may lead to other unintended side-effects to other users. 🤷‍♂️

Possibly interesting but not helpful since `num_ctx` isn't in the list, LiteLLM supposedly translates input parameters to compatible ones on the output side: https://litellm.vercel.app/docs/completion/input

**Reply by flyfox666** [reply]
Yes, thank you very much for the quick reply, it makes a lot more sense that this issue arose when look at it that way. I'm guessing there's been a recent update to the api backend of Google's gemini with more restrictions. Of course, perhaps, maybe it needs to buy a membership for google ai studio, after all, google has been emailing me lately to tell me that the free service is about to be updated :(

Looking at litellm's documentation, it does seem that there is no listing for num_ctx, I may have to go back and post an issue to ask about that.

Also, I'm going to open a new issue about embedding, I found that docker gives an error when changing the value of RAG_EMBEDDING_MODEL_DEVICE_TYPE in the env environment variable from cpu to gpu, and it doesn't seem to support the change with the default model:all-MiniLM-L6-v2. However I changed the embedding to intfloat/multilingual-e5-large, gpu mode is ok.

Also, would webui update for matching the ollama new update for  embedding?
https://ollama.com/blog/embedding-models

**Reply by ghost** [reply]
I wonder if this is related to recent cost map changes, looks like two Gemini Pro variants were just added: BerriAI/litellm@7f137bc9092875041f9513f3cbe2ff66246d2859. You may also want to ensure that you're setting this environment variable if you plan to use the cloud models with LiteLLM: `--env 'LITELLM_LOCAL_MODEL_COST_MAP=False'`

Edit: Also, `RAG_EMBEDDING_MODEL_DEVICE_TYPE` has been replaced with `DEVICE_TYPE` per open-webui/open-webui#1165 in **v0.1.118**. I think that `USE_CUDA_DOCKER=True` is currently the only way to change from `cpu` to `cuda`.

**Reply by flyfox666** [reply]
About this issue I also tried to solve it, but it is an imperfect way to handle it, after executing the cuda version of open webui docker, I traced the docker run with the following error: Traceback (most recent call last): File "/usr/local/lib/ python3.11/site-packages/litellm/llms/gemini.py", line 177, in async_streaming generation_config=genai.types.GenerationConfig( TypeError : GenerationConfig.__init__() got an unexpected keyword argument 'num_ctx'

Then I found the file /usr/local/lib/python3.11/site-packages/litellm/llms/gemini.py, and after asking gpt4, I added the following code before ## LOGGING in this file:

## Remove unsupported num_ctx arguments
inference_params.pop('num_ctx', None)         # Safely remove num_ctx from the dictionary

After executing gemini it doesn't report an error, and as @justinh-rahb said, the num_ctx parameter can't be transferred to google.

One thing I'm struggling with though is that I can only access this file directory in docker desktop, but I can't see it in wsl, and maybe because I'm a hobbyist (my main job is in finance) and I'm not very familiar with the code, I don't know if this file is included in your code.

The version I have installed now is: ghcr.io/open-webui/open-webui:cuda

Thank you again, the experience with openwebui is amazing!

**Reply by justinh-rahb** [reply]
@flyfox666 tbh that seems like a bug to be fixed that should be filed with the LiteLLM project.

**Reply by flyfox666** [reply]
let me try to make a issue in litellm

**Reply by cheahjs** [reply]
Haven't looked into it too hard, but we have this workaround on OpenAI calls that was introduced in https://github.com/open-webui/open-webui/pull/927. This could be harmonised with how LiteLLM is called to resolve the issue. https://github.com/open-webui/open-webui/blob/0399a69b73de9789c4221acedea70d528e1346c4/backend/apps/openai/main.py#L290-L295

**Reply by flyfox666** [reply]
@cheahjs Thanks for the tip!

**Comment by flyfox666** [comment]
> I wonder if this is related to recent cost map changes, looks like two Gemini Pro variants were just added: [BerriAI/litellm@7f137bc](https://github.com/BerriAI/litellm/commit/7f137bc9092875041f9513f3cbe2ff66246d2859). You may also want to ensure that you're setting this environment variable if you plan to use the cloud models with LiteLLM: `--env 'LITELLM_LOCAL_MODEL_COST_MAP=False'`
> 
> Edit: Also, `RAG_EMBEDDING_MODEL_DEVICE_TYPE` has been replaced with `DEVICE_TYPE` per #1165 in **v0.1.118**. I think that `USE_CUDA_DOCKER=True` is currently the only way to change from `cpu` to `cuda`.

HI 
Thanks very much, have just tried setting LITELLM_LOCAL_MODEL_COST_MAP=False , and also looked at Litellm's documentation, the problem with num_ctx reporting an error still exists, I will follow up on that. The main reason I changed the context length was to make some of the local model answers a little richer. Again, thank you very much for the heads up.

Also, is it possible to set it to use gpu by directly modifying USE_CUDA_DOCKER=True? My ollama webui installation on docker desktop for windows. I can not find the DEVICE_TYPE env,or I need to add the DEVICE_TYPE into env setting?

**Reply by justinh-rahb** [reply]
> Also, is it possible to set it to use gpu by directly modifying USE_CUDA_DOCKER=True? My ollama webui installation on docker desktop for windows. I can not find the DEVICE_TYPE env,or I need to add the DEVICE_TYPE into env setting?

You should just use the `:cuda` tagged Docker image now, it will have the device types set appropriately. As for changing embedding models, this is not to be done lightly, it will break any previously embedded documents and you must reset your vector DB.

**Reply by flyfox666** [reply]
yeap，I just use the :cuda tagged Docker , but with RuntimeError: Found no NVIDIA driver on your system. Please check that you have an NVIDIA GPU and installed a driver from http://www.nvidia.com/Download/index.aspx ;  all the cuda support I have set

I use the docker for windows desktop , I will try pure wsl2 enviroment later

**Reply by justinh-rahb** [reply]
@flyfox666 that is odd, it should _just work_ with Docker Desktop and Nvidia drivers installed for Windows, how did you invoke the docker container?

Our testing admittedly focused on native Linux. I have a colleague at work that could test a similar environment, I'll post results.

**Reply by flyfox666** [reply]
@justinh-rahb really thanks，I am now trying to downgrade the cuda tookit version .  I use the docker container in docker desktop at first , now i tred the pure wsl2 envirnment, but still error.

**Reply by ghost** [reply]
Docker Desktop GPU  support on Windodws actually requires WSL:
> [Currently GPU support in Docker Desktop is only available on Windows with the WSL2 backend.](https://docs.docker.com/desktop/gpu/)

I was able to get it working under WSL & Docker with our :cuda image but not with Podman & CDI. You'll need to ensure that WSL is working with CUDA and that Docker Desktop is configured to use the WSL backend. Relevant documentation:
https://docs.nvidia.com/cuda/wsl-user-guide/index.html

Good luck!

**Reply by flyfox666** [reply]
hi @lainedfles thank you very much for your answer, 

I followed exactly what you said and reinstalled wsl and docker desktop, now it works perfectly, the main reason I did this is because I changed the embedding model to:[multilingual-e5-large], the cpu runs this model! The main reason I did this is because I changed the embedding model to :[multilingual-e5-large], the cpu was running this model and it was lagging a bit, now it's running completely on the gpu. Thank you very much. One of the very tricky things is to make sure to follow exactly what is said in https://docs.nvidia.com/cuda/wsl-user-guide/index.html.

One thing to note: the docker run command needs to be executed as follows, be sure to add the command: --gpus all

docker run -d -p xxxx:8080 --gpus all --add-host=host.docker.internal:host-gateway -v xxxx -webui:/app/backend/data --name open-webui --restart always ghcr.io/open-webui/open-webui:cuda

Otherwise, it will display an unavailable gpu error


Also thanks to @justinh-rahb @tjbck

**Comment by ghost** [comment]
> HI Thanks very much, have just tried setting LITELLM_LOCAL_MODEL_COST_MAP=False , and also looked at Litellm's documentation, the problem with num_ctx reporting an error still exists, I will follow up on that. The main reason I changed the context length was to make some of the local model answers a little richer. Again, thank you very much for the heads up.

I haven't read the code but it could be that the cost map only affects addition of new/updated models. Perhaps when adding an unrecognized model, it defaulted to Ollama somehow. If your cost map is always up to date, you may not have a problem in the future when adding models supported by LiteLLM upstream.

A valid test may be to:
1. Remove the model from LiteLLM
2. Restart the container with the cost map update enabled
3. Re-add the same model

> Also, is it possible to set it to use gpu by directly modifying USE_CUDA_DOCKER=True? My ollama webui installation on docker desktop for windows. I can not find the DEVICE_TYPE env,or I need to add the DEVICE_TYPE into env setting?

I don't think it's possible but worth a try. If the default library dependencies included in your image are enough for your GPU, it might work. However I doubt it. A better option is to download one of the newly released `:cuda` variant containers or build yourself with the `--build-arg="USE_CUDA=true"` argument. This ensures that all of the necessary dependencies are included in the image. Note, the CUDA images are significantly larger (7GB+)! See [the published packages](https://github.com/open-webui/open-webui/pkgs/container/open-webui).

**Reply by flyfox666** [reply]
A valid test may be to:

Remove the model from LiteLLM
Restart the container with the cost map update enabled
Re-add the same model

Just did the test and it still doesn't work, the problem seems to exist with the page cache at once, clearing the cookie brought it back to normal. Thank you very much for your thoughts!

----------------------------------------------------------------------------------------------------------------------------------------------------
USE_CUDA_DOCKER=True
Container won't start after setup, it says missing dependencies.

I'll try installing cuda under WSL2 on windows 11 after I'm done and see if that fixes it.

**Comment by flyfox666** [comment]
Already bug report in LiteLLM project
https://github.com/BerriAI/litellm/issues/2987
